src/main/java/bio/terra/service/dataset/flight/UnlockDatasetStep.java

```diff
@@ -3,40 +3,58 @@
 import bio.terra.common.exception.RetryQueryException;
 import bio.terra.service.dataset.DatasetDao;
 import bio.terra.service.dataset.exception.DatasetLockException;
+import bio.terra.service.dataset.flight.ingest.OptionalStep;
 import bio.terra.stairway.FlightContext;
-import bio.terra.stairway.Step;
+import bio.terra.stairway.FlightMap;
 import bio.terra.stairway.StepResult;
 import bio.terra.stairway.StepStatus;
 import java.util.UUID;
+import java.util.function.Predicate;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class UnlockDatasetStep implements Step {
+public class UnlockDatasetStep extends OptionalStep {
 
   private static Logger logger = LoggerFactory.getLogger(UnlockDatasetStep.class);
 
   private final DatasetDao datasetDao;
-  private UUID datasetId;
   private boolean sharedLock; // default to false
+  private UUID datasetId;
 
-  public UnlockDatasetStep(DatasetDao datasetDao, UUID datasetId, boolean sharedLock) {
+  public UnlockDatasetStep(
+      DatasetDao datasetDao,
+      UUID datasetId,
+      boolean sharedLock,
+      Predicate<FlightContext> doCondition) {
+    super(doCondition);
     this.datasetDao = datasetDao;
     this.datasetId = datasetId;
 
     // this will be set to true for a shared lock, false for an exclusive lock
     this.sharedLock = sharedLock;
   }
 
+  public UnlockDatasetStep(DatasetDao datasetDao, UUID datasetId, boolean sharedLock) {
+    this(datasetDao, datasetId, sharedLock, OptionalStep::alwaysDo);
+  }
+
   public UnlockDatasetStep(DatasetDao datasetDao, boolean sharedLock) {
-    this(datasetDao, null, sharedLock);
+    this(datasetDao, null, sharedLock, OptionalStep::alwaysDo);
+  }
+
+  public UnlockDatasetStep(
+      DatasetDao datasetDao, boolean sharedLock, Predicate<FlightContext> doCondition) {
+    this(datasetDao, null, sharedLock, doCondition);
   }
 
   @Override
-  public StepResult doStep(FlightContext context) {
-    // In the create case, we won't have the dataset id at step creation. We'll expect it to be in
-    // the working map.
+  public StepResult doOptionalStep(FlightContext context) {
     if (datasetId == null) {
-      datasetId = context.getWorkingMap().get(DatasetWorkingMapKeys.DATASET_ID, UUID.class);
+      FlightMap map = context.getWorkingMap();
+
+      // In the create case, we won't have the dataset id at step creation. We'll expect it to be in
+      // the working map.
+      datasetId = map.get(DatasetWorkingMapKeys.DATASET_ID, UUID.class);
```

Comment:
Since we're already wrapping the step logic in a `try` block, we could use `Objects.requireNonNullElse` to do the null-checking for us, and tack another `catch` to turn the NPE into a failing `StepResult`. 

```
  @Override
  public StepResult doOptionalStep(FlightContext context) {
    FlightMap map = context.getWorkingMap();
    boolean rowUpdated;
    try {
      // In the create case, we won't have the dataset id at step creation. We'll expect it to be in
      // the working map.
      datasetId =
          Objects.requireNonNullElse(
              datasetId, map.get(DatasetWorkingMapKeys.DATASET_ID, UUID.class));
      if (sharedLock) {
        rowUpdated = datasetDao.unlockShared(datasetId, context.getFlightId());
      } else {
        rowUpdated = datasetDao.unlockExclusive(datasetId, context.getFlightId());
      }
      logger.debug("rowUpdated on unlock = " + rowUpdated);
    } catch (RetryQueryException retryQueryException) {
      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY);
    } catch (NullPointerException ex) {
      return new StepResult(
          StepStatus.STEP_RESULT_FAILURE_FATAL,
          new DatasetLockException("Expected dataset id in working map."));
    }
    return StepResult.getStepResultSuccess();
  }
  ```

Reply:
We don't require every reference of the unlockDatasetStep to provide a datasetId - So, this is basically catching the case that we pass in null and grab it from the flight.

Reply:
Yup! I just think this makes a little more sense rather than having nested `if` statements and reuses the `try` block from the row update.

Reply:
Got it - that is cleaner! Thanks.